### ISSUE TITLE
Implicit arena allocation resulting in OOM on Linux

### ISSUE THREAD

**Issue by ekigwana** [google/protobuf#4270]
I do not explicitly use arena allocation but keep getting OOM with resident memory use at less than 50M. It is clean as that the virtual memory allocations are being performed by new_heap (arean.c:486) and new_heap(arena.c:497) via arena_get2_part_4(arena.c:703).

I think the implicit arena use is here because when this data is not available this code is not called and the VIRT does not grow to the point of an OOM:

```
std::shared_ptr<pb::hw::pdu::rcurx::RCURX> yaml_to_pb(std::shared_ptr<std::string> req)
{
auto json_str = std::make_shared<std::string>(YAMLtoJSON(req));
auto rep = std::make_shared<pb::hw::pdu::rcurx::RCURX>();
auto data = rep->mutable_status();
auto status = google::protobuf::util::JsonStringToMessage(*json_str, data);

if (status.error_code() != google::protobuf::util::error::OK)
   std::cerr << "Error - " << status.error_message() << std::endl;
   data->set_ret_code(std::make_error_code(std::errc::state_not_recoverable).value());
}

return rep;
```

I have observed the same behavior with protobuf-3.4.1 and protobuf-3.5.1.
I have attached the massif output to show the allocation growth over time.

In the mean time, is there a way for me to just disable arena allocation globally?

[massif.out.20748.txt](https://github.com/google/protobuf/files/1690353/massif.out.20748.txt)

**Comment by xfxyjwf** [comment]
@ekigwana Are you able to figure the root cause of the problem?

**Comment by ekigwana** [comment]
Sorry for the delay in replying but I wanted to be sure I had diagnosed the issue correctly.

Turns out it was pure coincidence that protobufs were involved (I use them a lot).

Looked at the OOM issue and found that i needed to adjust the kernel configuration.

Previous option caused task to run out of virtual memory due to glibc's performance tuning that allocates memory per thread:
Enabling Transparent Hugepage always, can increase the
memory footprint of applications without a guaranteed
benefit but it will work automatically for all applications.

New option (which was used in previous kernels. Not sure how it was dropped but was not deliberate in any case:
Enabling Transparent Hugepage madvise, will only provide a
performance improvement benefit to the applications using 
madvise(MADV_HUGEPAGE) but it won't risk to increase the
memory footprint of applications without a guaranteed 
benefit.

I also set a hard limit on the maximum number of arenas that can be created.
`mallopt(M_ARENA_MAX, 8);`

The combination reduced the virtual memory usage an order of magnitude to Mega instead of Giga for our massively threaded server that uses protobufs everywhere. So uptime when from randomly between 3 hours and 11 to now It looks like it should stay up forever.

Not sure I needed to set M_ARENA_MAX but I did since I was up against a deadline so testing is key. I got higher CPU use but I can accept that due to the deadline since I don't time to twiddle at this time.
